File: LeetCode problem set/13_Roman_to_Integer.cpp
// https://leetcode.com/problems/roman-to-integer/
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
  map<char, int> roman_values = {
      {'I', 1},
      {'V', 5},
      {'X', 10},
      {'L', 50},
      {'C', 100},
      {'D', 500},
      {'M', 1000}
  };
  int romanToInt(string s) {
    int count = 0;
    int i = 0;
    while(i < s.length()) {
      int num = 0;
      if(roman_values[s[i]] >= roman_values[s[i+1]]) {
        num = roman_values[s[i]];
        i++;
      } else {
        num = roman_values[s[i+1]] - roman_values[s[i]];
        i += 2;
      }
      count += num;
    }
    return count;
  };
};

int main() {
  ios_base::sync_with_stdio(0), cin.tie(0), cout.tie(0);
  string s = "IV";
  // Solution().romanToInt(s);
   cout << Solution().romanToInt(s) << "\n";
}

/*
Example 1:
------------
Input: s = "III"
Output: 3
Explanation: III = 3.

Example 2:
------------
Input: s = "LVIII"
Output: 58
Explanation: L = 50, V= 5, III = 3.

Example 3:
-----------
Input: s = "MCMXCIV"
Output: 1994
Explanation: M = 1000, CM = 900, XC = 90 and IV = 4.
*/

File: LeetCode problem set/13_Roman_to_Integer.js
// https://leetcode.com/problems/roman-to-integer/

const romanVal = {
  I: 1,
  V: 5,
  X: 10,
  L: 50,
  C: 100,
  D: 500,
  M: 1000,
};

const romanToInt = (s) => {
  let count = 0;

  for (let i = 0; i < s.length; i++) {
    let cur = s[i];
    let next = s[i + 1];
    if (romanVal[cur] < romanVal[next]) {
      count -= romanVal[cur];
    } else {
      count += romanVal[cur];
    }
  }
  return count;
};

let roman = romanToInt("LVIII");
console.log("🚀 ~ file: 13_Roman_to_Integer.js:29 ~ romanToInt:", roman);
